File: backend/internal/handler/clientes.go
package handler

import (
	"net/http"
	"strconv"

	"blendpos/internal/apierror"
	"blendpos/internal/dto"
	"blendpos/internal/middleware"
	"blendpos/internal/service"

	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)

type ClientesHandler struct{ svc service.ClienteService }

func NewClientesHandler(svc service.ClienteService) *ClientesHandler {
	return &ClientesHandler{svc: svc}
}

// Crear godoc
// @Summary      Crear cliente
// @Description  Crea un nuevo cliente con cuenta corriente (fiado).
// @Tags         clientes
// @Accept       json
// @Produce      json
// @Security     BearerAuth
// @Param        body body dto.CrearClienteRequest true "Datos del cliente"
// @Success      201  {object} dto.ClienteResponse
// @Failure      400  {object} apierror.APIError
// @Router       /v1/clientes [post]
func (h *ClientesHandler) Crear(c *gin.Context) {
	var req dto.CrearClienteRequest
	if !bindAndValidate(c, &req) {
		return
	}
	resp, err := h.svc.Crear(c.Request.Context(), req)
	if err != nil {
		c.JSON(http.StatusBadRequest, apierror.New(err.Error()))
		return
	}
	id, _ := uuid.Parse(resp.ID)
	middleware.AuditLog(c, "create", "cliente", &id, map[string]interface{}{"nombre": req.Nombre})
	c.JSON(http.StatusCreated, resp)
}

// Listar godoc
// @Summary      Listar clientes
// @Description  Lista clientes activos con búsqueda opcional por nombre.
// @Tags         clientes
// @Produce      json
// @Security     BearerAuth
// @Param        search query string false "Buscar por nombre"
// @Param        page   query int    false "Página (default 1)"
// @Param        limit  query int    false "Registros por página (default 50)"
// @Success      200    {object} dto.ClienteListResponse
// @Router       /v1/clientes [get]
func (h *ClientesHandler) Listar(c *gin.Context) {
	search := c.Query("search")
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "50"))

	resp, err := h.svc.Listar(c.Request.Context(), search, page, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, apierror.New("Error al listar clientes"))
		return
	}
	c.JSON(http.StatusOK, resp)
}

// ObtenerPorID godoc
// @Summary      Obtener detalle de cliente
// @Description  Retorna el cliente con su saldo deudor actual.
// @Tags         clientes
// @Produce      json
// @Security     BearerAuth
// @Param        id path string true "UUID del cliente"
// @Success      200 {object} dto.ClienteResponse
// @Failure      400 {object} apierror.APIError
// @Router       /v1/clientes/{id} [get]
func (h *ClientesHandler) ObtenerPorID(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, apierror.New("ID invalido"))
		return
	}
	resp, err := h.svc.ObtenerPorID(c.Request.Context(), id)
	if err != nil {
		c.JSON(http.StatusNotFound, apierror.New(err.Error()))
		return
	}
	c.JSON(http.StatusOK, resp)
}

// Actualizar godoc
// @Summary      Actualizar cliente
// @Description  Actualiza datos del cliente (nombre, teléfono, límite crédito, etc.)
// @Tags         clientes
// @Accept       json
// @Produce      json
// @Security     BearerAuth
// @Param        id   path string true "UUID del cliente"
// @Param        body body dto.UpdateClienteRequest true "Campos a actualizar"
// @Success      200 {object} dto.ClienteResponse
// @Failure      400 {object} apierror.APIError
// @Router       /v1/clientes/{id} [put]
func (h *ClientesHandler) Actualizar(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, apierror.New("ID invalido"))
		return
	}
	var req dto.UpdateClienteRequest
	if !bindAndValidate(c, &req) {
		return
	}
	resp, err := h.svc.Actualizar(c.Request.Context(), id, req)
	if err != nil {
		c.JSON(http.StatusBadRequest, apierror.New(err.Error()))
		return
	}
	middleware.AuditLog(c, "update", "cliente", &id, nil)
	c.JSON(http.StatusOK, resp)
}

// RegistrarPago godoc
// @Summary      Registrar pago de cuenta corriente
// @Description  Registra un pago que reduce el saldo deudor del cliente.
// @Tags         clientes
// @Accept       json
// @Produce      json
// @Security     BearerAuth
// @Param        id   path string true "UUID del cliente"
// @Param        body body dto.RegistrarPagoClienteRequest true "Monto y descripción"
// @Success      201 {object} dto.MovimientoCuentaResponse
// @Failure      400 {object} apierror.APIError
// @Router       /v1/clientes/{id}/pago [post]
func (h *ClientesHandler) RegistrarPago(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, apierror.New("ID invalido"))
		return
	}
	var req dto.RegistrarPagoClienteRequest
	if !bindAndValidate(c, &req) {
		return
	}
	resp, err := h.svc.RegistrarPago(c.Request.Context(), id, req)
	if err != nil {
		c.JSON(http.StatusBadRequest, apierror.New(err.Error()))
		return
	}
	middleware.AuditLog(c, "pago", "cliente", &id, map[string]interface{}{"monto": req.Monto.StringFixed(2)})
	c.JSON(http.StatusCreated, resp)
}

// ListarMovimientos godoc
// @Summary      Historial de movimientos de cuenta corriente
// @Description  Lista paginada de movimientos (cargos, pagos, ajustes) del cliente.
// @Tags         clientes
// @Produce      json
// @Security     BearerAuth
// @Param        id    path  string true  "UUID del cliente"
// @Param        page  query int    false "Página (default 1)"
// @Param        limit query int    false "Registros por página (default 50)"
// @Success      200 {object} dto.MovimientosListResponse
// @Router       /v1/clientes/{id}/movimientos [get]
func (h *ClientesHandler) ListarMovimientos(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		c.JSON(http.StatusBadRequest, apierror.New("ID invalido"))
		return
	}
	page, _ := strconv.Atoi(c.DefaultQuery("page", "1"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "50"))

	resp, err := h.svc.GetMovimientos(c.Request.Context(), id, page, limit)
	if err != nil {
		c.JSON(http.StatusInternalServerError, apierror.New("Error al listar movimientos"))
		return
	}
	c.JSON(http.StatusOK, resp)
}

// ListarDeudores godoc
// @Summary      Listar deudores
// @Description  Lista todos los clientes con saldo deudor > 0, ordenados por saldo descendente.
// @Tags         clientes
// @Produce      json
// @Security     BearerAuth
// @Success      200 {object} dto.ListDeudoresResponse
// @Router       /v1/clientes/deudores [get]
func (h *ClientesHandler) ListarDeudores(c *gin.Context) {
	resp, err := h.svc.GetDeudores(c.Request.Context())
	if err != nil {
		c.JSON(http.StatusInternalServerError, apierror.New("Error al listar deudores"))
		return
	}
	c.JSON(http.StatusOK, resp)
}

File: frontend/src/services/api/clientes.ts
// ─────────────────────────────────────────────────────────────────────────────
// Clientes API — Fiado / Cuenta Corriente
// POST /v1/clientes, GET /v1/clientes, GET /v1/clientes/:id,
// PUT /v1/clientes/:id, POST /v1/clientes/:id/pago,
// GET /v1/clientes/:id/movimientos, GET /v1/clientes/deudores
// ─────────────────────────────────────────────────────────────────────────────

import { apiClient } from '../../api/client';

// ── Response Types ────────────────────────────────────────────────────────────

export interface ClienteResponse {
    id: string;
    nombre: string;
    telefono?: string;
    email?: string;
    dni?: string;
    limite_credito: number;
    saldo_deudor: number;
    credito_disponible: number;
    activo: boolean;
    notas?: string;
    created_at: string;
    updated_at: string;
}

export interface ClienteListResponse {
    data: ClienteResponse[];
    total: number;
}

export interface MovimientoCuentaResponse {
    id: string;
    cliente_id: string;
    tipo: 'cargo' | 'pago' | 'ajuste';
    monto: number;
    saldo_posterior: number;
    referencia_id?: string;
    referencia_tipo?: string;
    descripcion?: string;
    created_at: string;
}

export interface MovimientosListResponse {
    data: MovimientoCuentaResponse[];
    total: number;
    page: number;
    limit: number;
}

export interface DeudorResponse {
    id: string;
    nombre: string;
    telefono?: string;
    saldo_deudor: number;
    limite_credito: number;
}

export interface ListDeudoresResponse {
    data: DeudorResponse[];
    total: number;
}

// ── Request Types ─────────────────────────────────────────────────────────────

export interface CrearClienteRequest {
    nombre: string;
    telefono?: string;
    email?: string;
    dni?: string;
    limite_credito: number;
    notas?: string;
}

export interface UpdateClienteRequest {
    nombre?: string;
    telefono?: string;
    email?: string;
    dni?: string;
    limite_credito?: number;
    activo?: boolean;
    notas?: string;
}

export interface RegistrarPagoRequest {
    monto: number;
    descripcion?: string;
}

// ── Decimal coercion helpers ──────────────────────────────────────────────────
// shopspring/decimal may serialize as quoted strings ("1500.00") or numbers
// depending on backend config. These helpers ensure the frontend always has
// proper numeric types regardless of backend serialization format.

function toNum(v: unknown): number {
    if (typeof v === 'number') return v;
    if (typeof v === 'string') return parseFloat(v) || 0;
    return 0;
}

function coerceCliente(raw: ClienteResponse): ClienteResponse {
    return {
        ...raw,
        limite_credito: toNum(raw.limite_credito),
        saldo_deudor: toNum(raw.saldo_deudor),
        credito_disponible: toNum(raw.credito_disponible),
    };
}

function coerceMovimiento(raw: MovimientoCuentaResponse): MovimientoCuentaResponse {
    return {
        ...raw,
        monto: toNum(raw.monto),
        saldo_posterior: toNum(raw.saldo_posterior),
    };
}

function coerceDeudor(raw: DeudorResponse): DeudorResponse {
    return {
        ...raw,
        saldo_deudor: toNum(raw.saldo_deudor),
        limite_credito: toNum(raw.limite_credito),
    };
}

// ── API Calls ─────────────────────────────────────────────────────────────────

/**
 * POST /v1/clientes  (supervisor, administrador)
 * Crea un nuevo cliente con cuenta corriente.
 */
export async function crearCliente(data: CrearClienteRequest): Promise<ClienteResponse> {
    const raw = await apiClient.post<ClienteResponse>('/v1/clientes', data);
    return coerceCliente(raw);
}

/**
 * GET /v1/clientes  (cajero, supervisor, administrador)
 * Lista clientes activos con busqueda opcional por nombre.
 */
export async function listarClientes(params?: {
    search?: string;
    page?: number;
    limit?: number;
}): Promise<ClienteListResponse> {
    const raw = await apiClient.get<ClienteListResponse>('/v1/clientes', {
        search: params?.search,
        page: params?.page ?? 1,
        limit: params?.limit ?? 50,
    });
    return { ...raw, data: raw.data.map(coerceCliente) };
}

/**
 * GET /v1/clientes/:id  (cajero, supervisor, administrador)
 * Detalle del cliente con saldo actual.
 */
export async function obtenerCliente(id: string): Promise<ClienteResponse> {
    const raw = await apiClient.get<ClienteResponse>(`/v1/clientes/${id}`);
    return coerceCliente(raw);
}

/**
 * PUT /v1/clientes/:id  (supervisor, administrador)
 * Actualiza datos del cliente.
 */
export async function actualizarCliente(id: string, data: UpdateClienteRequest): Promise<ClienteResponse> {
    const raw = await apiClient.put<ClienteResponse>(`/v1/clientes/${id}`, data);
    return coerceCliente(raw);
}

/**
 * POST /v1/clientes/:id/pago  (supervisor, administrador)
 * Registra un pago que reduce el saldo deudor.
 */
export async function registrarPago(id: string, data: RegistrarPagoRequest): Promise<MovimientoCuentaResponse> {
    const raw = await apiClient.post<MovimientoCuentaResponse>(`/v1/clientes/${id}/pago`, data);
    return coerceMovimiento(raw);
}

/**
 * GET /v1/clientes/:id/movimientos  (cajero, supervisor, administrador)
 * Historial paginado de movimientos de cuenta corriente.
 */
export async function listarMovimientos(
    id: string,
    params?: { page?: number; limit?: number },
): Promise<MovimientosListResponse> {
    const raw = await apiClient.get<MovimientosListResponse>(`/v1/clientes/${id}/movimientos`, {
        page: params?.page ?? 1,
        limit: params?.limit ?? 50,
    });
    return { ...raw, data: raw.data.map(coerceMovimiento) };
}

/**
 * GET /v1/clientes/deudores  (cajero, supervisor, administrador)
 * Lista todos los clientes con saldo > 0, ordenados por deuda descendente.
 */
export async function listarDeudores(): Promise<ListDeudoresResponse> {
    const raw = await apiClient.get<ListDeudoresResponse>('/v1/clientes/deudores');
    return { ...raw, data: raw.data.map(coerceDeudor) };
}
